crates/wasmtime/src/store.rs

```diff
@@ -1052,10 +1088,33 @@ impl<T> StoreInner<T> {
     }
 
     pub fn call_hook(&mut self, s: CallHook) -> Result<(), Trap> {
-        if let Some(hook) = &mut self.call_hook {
-            hook(&mut self.data, s)
+        // Need to borrow async_cx before the mut borrow of the limiter.
+        // self.async_cx() panicks when used with a non-async store, so
+        // wrap this in an option.
+        #[cfg(feature = "async")]
+        let async_cx = if self.inner.async_support() {
+            Some(self.inner.async_cx())
         } else {
-            Ok(())
+            None
+        };
+
+        match self.call_hook {
+            Some(CallHookInner::Sync(ref mut hook)) => hook(&mut self.data, s),
+
+            #[cfg(feature = "async")]
+            Some(CallHookInner::Async(ref mut handler)) => {
+                match async_cx.expect("CallHookInner requires async Store") {
+                    None => Err(Trap::new(
+                        "ignoring attempt to resume during fiber shutdown",
+                    )),
```

Comment:
Instead of returning a trap could this return `Ok(())`, perhaps with an assert that this is `CallHook::ExitingWasm` (I forget the precise name) since I think we should be guaranteed that the hook is only used in that one direction.

Additionally when I stop to think about this I think it's a small preexisting bug today where when a fiber is being torn down we shouldn't execute any user-provided code on the fiber, instead only destructors. Right now though if you have a synchronous call-hook installed then we'll still run the synchronous call hook when exiting wasm.

To handle that, perhaps my above comment about moving `self.inner.async_cx()` down here isn't so applicable. Perhaps if async support is enabled and if the cx isn't available then `Ok(())` is unconditionally returned to avoid calling any hook infrastructure? (also having the assert about `s: CallHook` I think)